app/upgrades/v15/upgrades.go

```diff
@@ -21,7 +44,243 @@ func CreateUpgradeHandler(
 		return vm, err
 	}
 }
+
+// UpgradeMinCommissionRate sets the minimum commission rate staking parameter to 5%
+// and updates the commission rate for all validators that have a commission rate less than 5%
+func UpgradeMinCommissionRate(ctx sdk.Context, sk stakingkeeper.Keeper) {
+	params := sk.GetParams(ctx)
+	params.MinCommissionRate = sdk.NewDecWithPrec(5, 2)
+	err := sk.SetParams(ctx, params)
+	if err != nil {
+		panic(err)
+	}
+
+	for _, val := range sk.GetAllValidators(ctx) {
+		if val.Commission.CommissionRates.Rate.LT(sdk.NewDecWithPrec(5, 2)) {
+			// set the commission rate to 5%
+			val.Commission.CommissionRates.Rate = sdk.NewDecWithPrec(5, 2)
+			// set the max rate to 5% if it is less than 5%
+			if val.Commission.CommissionRates.MaxRate.LT(sdk.NewDecWithPrec(5, 2)) {
+				val.Commission.CommissionRates.MaxRate = sdk.NewDecWithPrec(5, 2)
+			}
+			val.Commission.UpdateTime = ctx.BlockHeader().Time
+			sk.SetValidator(ctx, val)
+		}
+	}
+}
+
+// UpgradeSigningInfos updates the signing infos of validators for which
+// the consensus address is missing
+func UpgradeSigningInfos(ctx sdk.Context, sk slashingkeeper.Keeper) {
+	signingInfos := []slashingtypes.ValidatorSigningInfo{}
+
+	// update consensus address in signing info
+	// using the store key of validators
+	sk.IterateValidatorSigningInfos(ctx, func(address sdk.ConsAddress, info slashingtypes.ValidatorSigningInfo) (stop bool) {
+		if info.Address == "" {
+			info.Address = address.String()
+			signingInfos = append(signingInfos, info)
+		}
+
+		return false
+	})
+
+	for _, si := range signingInfos {
+		addr, err := sdk.ConsAddressFromBech32(si.Address)
+		if err != nil {
+			panic(err)
+		}
+		sk.SetValidatorSigningInfo(ctx, addr, si)
+	}
+}
+
+// ClawbackVestingFunds transfers the unvested tokens from the given vesting account
+// to the community pool
+func ClawbackVestingFunds(ctx sdk.Context, address sdk.AccAddress, keepers *keepers.AppKeepers) {
+	ak := keepers.AccountKeeper
+	bk := keepers.BankKeeper
+	dk := keepers.DistrKeeper
+	sk := *keepers.StakingKeeper
+
+	// get target account
+	account := ak.GetAccount(ctx, address)
+
+	// verify that it's a vesting account type
+	vestAccount, ok := account.(*vesting.ContinuousVestingAccount)
+	if !ok {
+		panic(
+			sdkerrors.Wrap(
+				sdkerrors.ErrInvalidType,
+				"account with address %s isn't a vesting account",
+			),
+		)
+	}
+
+	// unbond all delegations from vesting account
+	err := forceUnbondAllDelegations(sk, bk, ctx, address)
+	if err != nil {
+		panic(err)
+	}
+
+	// transfers still vesting tokens of BondDenom to community pool
+	_, vestingCoins := vestAccount.GetVestingCoins(ctx.BlockTime()).Find(sk.BondDenom(ctx))
+	err = forceFundCommunityPool(
+		ak,
+		dk,
+		bk,
+		ctx,
+		vestingCoins,
+		address,
+		keepers.GetKey(banktypes.StoreKey),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	// overwrite vesting account using its embedded base account
+	ak.SetAccount(ctx, vestAccount.BaseAccount)
```

Comment:
Can you explain why this needs to be `vestAccount.BaseAccount` and not just the `vestAccount`?

Reply:
Mainly, bc `vestAccount` is a `ContinuousVestingAccount` type which isn't required after migration since there will only remain unlocked and vested coins in the account.
Therefore, its embedded `BaseAccount` field that holds the basic account value is sufficient. 

The other option would have required updating the vesting account's original amount and end time, since it's lazy and calculates the vesting/vested coins according to the current block time rather than storing it.